src/data/fin.lean

```diff
@@ -1012,6 +1012,168 @@ def append {α : Type*} {o : ℕ} (ho : o = m + n) (u : fin m → α) (v : fin n
   then u ⟨i, h⟩
   else v ⟨(i : ℕ) - m, (nat.sub_lt_left_iff_lt_add (le_of_not_lt h)).2 (ho ▸ i.property)⟩
 
+variables {β : Type*} {o : ℕ} {u : fin m → β} {v : fin n → β}
+
+lemma append_apply_left (ho : o = m + n) (k : fin o) (h : (k : ℕ) < m) :
+  append ho u v k = u (k.cast_lt h) :=
+dif_pos h
+
+@[simp] lemma append_apply_left_cast (ho : o = m + n) (k : fin m) :
+  append ho u v (cast ho.symm (cast_add n k)) = u k :=
+(append_apply_left ho (cast ho.symm (cast_add n k)) k.2).trans $
+  congr_arg _ (fin.ext rfl)
+
+lemma append_apply_right (ho : o = m + n) (k : fin o) (h : ¬(k : ℕ) < m) :
+  append ho u v k = v ⟨(k : ℕ) - m, (nat.sub_lt_left_iff_lt_add $ le_of_not_lt h).2 (ho ▸ k.2)⟩ :=
+dif_neg h
+
+@[simp] lemma append_apply_right_cast (ho : o = m + n) (k : fin n) :
+  append ho u v (cast ho.symm (nat_add m k)) = v k :=
+(append_apply_right ho (cast ho.symm (nat_add m k)) (not_lt.2 $ le_add_right (le_refl _))).trans $
+  congr_arg _ (fin.ext $ nat.add_sub_cancel_left _ _)
+
+lemma append_nil {d : fin 0 → β} (ho : o = m) {x : fin o} :
+  @append 0 m _ o ho u d x = u (cast ho x) :=
+(@append_apply_left 0 m _ o _ _ ho x (ho ▸ x.2)).trans $ by congr
+
+@[simp] lemma append_nil' {d : fin 0 → β} :
+  append rfl u d = u :=
+funext $ λ x, (append_nil _).trans (congr_arg _ $ fin.ext rfl)
+
+lemma nil_append {d : fin 0 → β} (ho : o = m) {i : fin o} :
+  append (show o = 0 + m, by rwa zero_add) d u i = u (cast ho i) :=
+(append_apply_right _ _ (nat.not_lt_zero _)).trans $ by congr
+
+@[simp] lemma nil_append' {d : fin 0 → β} :
+  append (zero_add _).symm d u = u :=
+funext $ λ x, (nil_append rfl).trans (congr_arg _ $ fin.ext rfl)
+
+lemma append_one_eq_cons (ho : o = 1 + m) {x : β} (i : fin o) :
+  append ho (λ i : fin 1, x) u i = (fin.cons x u : fin m.succ → β)
+    (cast (ho.trans $ add_comm 1 m) i) :=
+begin
+  cases decidable.em (cast (ho.trans $ add_comm _ _) i = 0) with hy hy,
+  { rw [hy, cons_zero, append_apply_left _ i (show (i : ℕ) < 1, by
+          erw [(ext_iff _ _).1 hy, coe_zero]; exact succ_pos')] },
+  { rw [←fin.succ_pred (cast (ho.trans $ add_comm _ _) i) hy, cons_succ,
+        append_apply_right _ i (λ h, hy $ fin.ext $ le_antisymm
+          (nat.lt_succ_iff.1 h) (nat.zero_le _))],
+    congr },
+end
+
+lemma append_update_fst (ho : o = m + n) (i : fin m) (x : β) :
+  append ho (function.update u i x) v =
+    function.update (append ho u v) (cast_lt i $ ho.symm ▸ nat.lt_add_right i m n i.2) x :=
+begin
+  ext y,
+  cases decidable.em ((y : ℕ) < m),
+  { rw append_apply_left ho _ h,
+    rcases classical.em (y = cast ho.symm (cast_add n i)) with ⟨rfl, hi⟩,
+    { erw function.update_same,
+      convert function.update_same i _ _,
+      ext,
+      refl },
+    { rw [function.update_noteq, function.update_noteq, append_apply_left ho _ h],
+      all_goals { contrapose! h_1, try {rw h_1 <|> rw ←h_1}, ext, refl }}},
```

Comment:
This `h_1` is an auto-generated name, right? Could you explicitly name it? (Without opening in an editor, I'm guessing this is from the `cases` at the beginning?)

Reply:
In fact, I think the same is true for `h` above.

Reply:
And maybe in the following proof?